pkg/logs/adapter/minio.go

```diff
@@ -0,0 +1,188 @@
+package adapter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"github.com/minio/minio-go/v7"
+	"go.uber.org/zap"
+
+	"github.com/kubeshop/testkube/pkg/log"
+	"github.com/kubeshop/testkube/pkg/logs/events"
+	minioconnecter "github.com/kubeshop/testkube/pkg/storage/minio"
+)
+
+const (
+	defaultBufferSize = 1024 * 100 // 100KB
+	defaultWriteSize  = 1024 * 80  // 80KB
+)
+
+var _ Adapter = &MinioConsumer{}
+
+type ErrMinioConsumerDisconnected struct {
+}
+
+func (e ErrMinioConsumerDisconnected) Error() string {
+	return "minio consumer disconnected"
+}
+
+type BufferInfo struct {
+	Buffer *bytes.Buffer
+	Part   int
+}
+
+// MinioConsumer creates new MinioSubscriber which will send data to local MinIO bucket
+func NewMinioConsumer(endpoint, accessKeyID, secretAccessKey, region, token, bucket string, opts ...minioconnecter.Option) *MinioConsumer {
+	c := &MinioConsumer{
+		minioConnecter: minioconnecter.NewConnecter(endpoint, accessKeyID, secretAccessKey, region, token, bucket, log.DefaultLogger, opts...),
+		Log:            log.DefaultLogger,
+		bucket:         bucket,
+		region:         region,
+		disconnected:   false,
+		buffInfos:      make(map[string]BufferInfo),
+	}
+	minioClient, err := c.minioConnecter.GetClient()
+	if err != nil {
+		c.Log.Errorw("error connecting to minio", "err", err)
+		return c
+	}
+
+	c.minioClient = minioClient
+	exists, err := c.minioClient.BucketExists(context.TODO(), c.bucket)
+	if err != nil {
+		c.Log.Errorw("error checking if bucket exists", "err", err)
+		return c
+	}
+
+	if !exists {
+		err = c.minioClient.MakeBucket(context.TODO(), c.bucket,
+			minio.MakeBucketOptions{Region: c.region})
+		if err != nil {
+			c.Log.Errorw("error creating bucket", "err", err)
+			return c
+		}
+	}
+	return c
+}
+
+type MinioConsumer struct {
+	minioConnecter *minioconnecter.Connecter
+	minioClient    *minio.Client
+	bucket         string
+	region         string
+	Log            *zap.SugaredLogger
+	disconnected   bool
+	buffInfos      map[string]BufferInfo
+	mapLock        sync.RWMutex
+}
+
+func (s *MinioConsumer) Notify(id string, e events.Log) error {
+	if s.disconnected {
+		s.Log.Debugw("minio consumer disconnected", "id", id)
+		return ErrMinioConsumerDisconnected{}
+	}
+
+	if _, ok := s.GetBuffInfo(id); !ok {
+		s.UpdateBuffInfo(id, BufferInfo{Buffer: bytes.NewBuffer(make([]byte, 0, defaultBufferSize)), Part: 0})
+	}
+
+	chunckToAdd, err := json.Marshal(e)
+	if err != nil {
+		return err
+	}
+	chunckToAdd = append(chunckToAdd, []byte("\n")...)
+	buffInfo, _ := s.GetBuffInfo(id)
+	writer := buffInfo.Buffer
+	_, err = writer.Write(chunckToAdd)
+	if err != nil {
+		return err
+	}
+
+	if writer.Len() > defaultWriteSize {
+		buffInfo.Buffer = bytes.NewBuffer(make([]byte, 0, defaultBufferSize))
+		name := id + "-" + strconv.Itoa(buffInfo.Part)
+		buffInfo.Part++
+		s.UpdateBuffInfo(id, buffInfo)
+		go s.putData(name, writer)
```

Comment:
sounds like making it async will always return sucess from this method

Reply:
The idea was to make Notify highly available and not to block for a lot of time, @exu what do you think?